src/checkers/flashchecker.py

```diff
@@ -0,0 +1,118 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+import json
+import logging
+import urllib.error
+import urllib.request
+import urllib.parse
+
+from lib.externaldata import ExternalData, ExternalFile, Checker
+from lib import utils
+
+log = logging.getLogger(__name__)
+
+
+FLASH_BASE_URL = 'http://get.adobe.com/flashplayer/webservices/json/?{params}'
+
+FLATPAK_TO_FLASH_ARCH_MAP = {
+    'i386': 'x86-32',
+    'x86_64': 'x86-64',
+}
+
+BROWSER_TO_PAPI_MAP = {
+    'Chrome': 'ppapi',
+    'Firefox': 'npapi',
+}
+
+
+class FlashChecker(Checker):
+    def _should_check(self, external_data):
+        return external_data.checker_data.get('type') == 'flash'
+
+    def _get_arches(self, external_data):
+        if len(external_data.arches) != 1:
+            return None
+
+        flatpak_arch = external_data.arches[0]
+        return flatpak_arch, FLATPAK_TO_FLASH_ARCH_MAP.get(external_data.arches[0])
+
+    def check(self, external_data):
+        if not self._should_check(external_data):
+            log.debug('%s is not a flash type ext data', external_data.filename)
+            return
+
+        browser = external_data.checker_data['browser'].title()
+
+        try:
+            papi = BROWSER_TO_PAPI_MAP[browser]
+        except KeyError:
+            log.warning('%s has an invalid browser (should be Chrome or Firefox)',
+                        external_data.filename)
+            return
+
+        arches = self._get_arches(external_data)
+        if arches is None:
+            log.warning('%s has invalid only-arches (should be one of %s)',
+                        external_data.filename,
+                        ', '.join(f'[{arch}]' for arch in FLATPAK_TO_FLASH_ARCH_MAP))
+            return
+
+        flatpak_arch, flash_arch = arches
+
+        request_params = {'platform_type': 'Linux', 'eventname': 'flashplayerotherversions',
+                          'platform_arch': flash_arch}
+        request_url = FLASH_BASE_URL.format(params=urllib.parse.urlencode(request_params))
+        with urllib.request.urlopen(request_url) as resp:
+            latest_version_data = json.load(resp)
+
+        latest_version = None
+        latest_url = None
+
+        for version in latest_version_data:
+            if (version['installation_type'] == 'Standalone' and version['browser'] == browser
+                    and version['installer_architecture'] == flash_arch):
+                assert version['platform'] == 'Linux'
+
+                latest_version = version['Version']
+                latest_url = version['download_url']
+                break
+
+        if latest_version is None:
```

Comment:
I am keen on spelling this kind of loop as:

```python3
for version in latest_version_data:
    if ...:
         break
else:
    # this block runs if the break is never reached
```

however I know this is not a universally-popular Python feature :-)

Reply:
Ah too much C++ lately, forgot all about that.